src/iris/api.py

```diff
@@ -4243,18 +4243,68 @@ def on_get(self, req, resp):
             stats = app_stats.calculate_global_stats(connection, cursor, fields_filter=fields_filter)
 
         else:
-            cursor.execute('SELECT `statistic`, `value` FROM `global_stats`')
+            cursor.execute('SELECT `statistic`, `value`, `timestamp` FROM `global_stats`')
             if cursor.rowcount == 0:
                 logger.exception('Error retrieving global stats from db')
                 raise HTTPInternalServerError('Error retrieving global stats from db')
-            stats = {row[0]: row[1] for row in cursor}
+
+            stats = {}
+
+            for row in cursor:
+                # {statistic : {timestamp: value, timestamp: value}}
+                if stats.get(row[0]):
+                    stats[row[0]][row[2]] = row[1]
+                else:
+                    stats[row[0]] = {}
+                    stats[row[0]][row[2]] = row[1]
 
         cursor.close()
         connection.close()
         resp.status = HTTP_200
         resp.body = ujson.dumps(stats, sort_keys=True)
 
 
+class HighPriorityIncidents(object):
+    allow_read_no_auth = True
+
+    def __init__(self, config):
+        cfg = config.get('app-stats', {})
+        # Calculate stats in real time (True), or query offline stats
+        # generated by the app stats daemon (False)
+        self.real_time = cfg.get('real_time', True)
+
+    def on_get(self, req, resp):
+
+        connection = db.engine.raw_connection()
+        cursor = connection.cursor()
+        stats = {}
+        if self.real_time:
+            stats = app_stats.calculate_high_priority_incidents(connection, cursor)
+        else:
+            cursor.execute('SELECT timestamp, name, value FROM application_stats JOIN application on application_stats.application_id = application.id WHERE statistic = "high_priority_incidents_last_week" ORDER BY value DESC')
+            if cursor.rowcount == 0:
+                logger.exception('Error retrieving hpi stats from db')
+                raise HTTPInternalServerError('Error retrieving stats from db')
```

Comment:
If this raises, the connection and cursor close lines will never get called, leading to us potentially leaking connetions.

Also, why is a rowcount of 0 worthy of a 5xx? 5xx usually refers to things actually being broken / database exceptions being raised

Reply:
The reason it's raising a 5xx is because unless no apps exist in the database the rowcount should never be 0 here. If the rowcount is 0 it means that the app_stats script failed and the stats tables are in a bad state.